io.openems.edge.common/src/io/openems/edge/common/statemachine/StateMachine.java

```diff
@@ -0,0 +1,117 @@
+package io.openems.edge.common.statemachine;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.openems.common.exceptions.OpenemsError.OpenemsNamedException;
+
+/**
+ * Manages the States of the StateMachine.
+ * 
+ * @param <STATE>   the {@link State} type, e.g. typically an enum
+ * @param <CONTEXT> the context type, i.e. a class wrapping a State-Machine
+ *                  context
+ */
+public class StateMachine<STATE extends State<STATE, CONTEXT>, CONTEXT> {
+
+	private final Logger log = LoggerFactory.getLogger(StateMachine.class);
+
+	private final STATE initialState;
```

Comment:
Yes, because State can be reset to this State even later if an error happens. With Java generics I am not able to get the UNDEFINED state in another way (had tried it...)